dwm - add media key support for brightness and volume
This patch adds support for:
- Screen brightness control (F1/F2 keys, keycodes 232/233)  
- Volume control (F11/F12 keys, keycodes 122/123)
- Mute toggle (F10 key, keycode 121)

Requires: brightnessctl and pamixer


diff --git a/config.def.h b/config.def.h
index 149f00d..2068d7f 100644
--- a/config.def.h
+++ b/config.def.h
@@ -113,8 +113,20 @@ static char dmenumon[2] = "0"; /* component of dmenucmd, manipulated in spawn()
 //static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", col_gray1, "-nf", col_gray3, "-sb", col_cyan, "-sf", col_gray4, NULL };
 static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", normbgcolor, "-nf", normfgcolor, "-sb", selbordercolor, "-sf", selfgcolor, NULL };
 static const char *termcmd[]  = { "st", NULL };
+//2fa
 static const char *tfacmd[] = { "authenticator", "dmenu", NULL };
 
+//macbookpro display/kb brightness controls/volume
+static const char *brightup[]   = { "brightnessctl", "set", "+5%", NULL };
+static const char *brightdown[] = { "brightnessctl", "set", "5%-", NULL };
+static const char *kbdlightup[]   = { "brightnessctl", "--device=apple::kbd_backlight", "set", "+10%", NULL };
+static const char *kbdlightdown[] = { "brightnessctl", "--device=apple::kbd_backlight", "set", "10%-", NULL };
+static const char *volup[]   = { "pamixer", "--increase", "5", NULL };
+static const char *voldown[] = { "pamixer", "--decrease", "5", NULL };
+static const char *volmute[] = { "pamixer", "--toggle-mute", NULL };
+
+
+
 static const Key keys[] = {
 	/* modifier                     key        function        argument */
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
@@ -151,6 +163,7 @@ static const Key keys[] = {
 
     // 2fa
     { MODKEY,                       XK_e,      spawn,          {.v = tfacmd } },
+
 	{ MODKEY,                       XK_Tab,    view,           {0} },
 	{ MODKEY|ShiftMask,             XK_c,      killclient,     {0} },
 	{ MODKEY,                       XK_t,      setlayout,      {.v = &layouts[0]} },
@@ -165,6 +178,16 @@ static const Key keys[] = {
 	{ MODKEY|ShiftMask,             XK_comma,  tagmon,         {.i = -1 } },
 	{ MODKEY|ShiftMask,             XK_period, tagmon,         {.i = +1 } },
 	{ MODKEY,                       XK_F5,     xrdb,           {.v = NULL } },
+    //mbp m1max keys
+    { 0,                            XF86XK_LaunchA,   spawn, {.v = kbdlightup } },
+    { 0,                            XF86XK_Search, spawn, {.v = kbdlightdown } },
+
+    { 0,                            XF86XK_MonBrightnessUp,   spawn, {.v = brightup } },
+    { 0,                            XF86XK_MonBrightnessDown, spawn, {.v = brightdown } },
+    { 0,                            XF86XK_AudioRaiseVolume,  spawn, {.v = volup } },
+    { 0,                            XF86XK_AudioLowerVolume,  spawn, {.v = voldown } },
+    { 0,                            XF86XK_AudioMute,         spawn, {.v = volmute } },
+
 	TAGKEYS(                        XK_1,                      0)
 	TAGKEYS(                        XK_2,                      1)
 	TAGKEYS(                        XK_3,                      2)
diff --git a/dwm.c b/dwm.c
index 6534772..51e3b61 100644
--- a/dwm.c
+++ b/dwm.c
@@ -32,6 +32,7 @@
 #include <sys/wait.h>
 #include <X11/cursorfont.h>
 #include <X11/keysym.h>
+#include <X11/XF86keysym.h>
 #include <X11/Xatom.h>
 #include <X11/Xlib.h>
 #include <X11/Xproto.h>
